### ISSUE TITLE
Marksman rifle trail visually pierces even super fat walls it definitely can't deal any damage through

### ISSUE THREAD

**Issue by WhenTanksFly** [ReactiveDrop/reactivedrop_public_src#241]
Example: Operation Cleansweep - Landing Bay 7, the big pillar-ish thing right after the spawnpoint room.
![pew](https://user-images.githubusercontent.com/112336094/187431420-7874a651-ecf8-453b-8269-1f8b0cea6b38.jpg)
[sniper_trail_fail_demo.zip](https://github.com/ReactiveDrop/reactivedrop_public_src/files/9452588/sniper_trail_fail_demo.zip) (v7.1.3.5)

**Comment by jhh8** [comment]
blocked by engine label?

**Comment by jhh8** [comment]
solution could be making the particle something like 16 units long and spamming many particles in a row until a unpenetratable object is hit

**Comment by BenLubar** [comment]
I don't think we're blocked by the engine here. I'll have to look at how the marksman rifle tracer is implemented but piercing bullets are a fully non-engine-side implementation.

**Comment by jhh8** [comment]
i mean cutting off the particle half way through so its not shown, not sure if sniper should be able to pierce such fat walls

**Comment by BenLubar** [comment]
I'm pretty sure the logic for displaying tracers is also in client.dll and not the engine.

**Comment by anf3is** [comment]
tl;dr displaying works fine, but tracing bullet path and finding hit targets sometimes fails.

`sv_showimpact 1;` helps a lot.

---

`FirePenetratingBullets` and `MakeTracer` both seemingly work without a hunch.

But `AI_TraceLine` is buggy.

When a trace hits a world object, the following trace starts with a small offset, pushing it forward in an attempt to penetrate past the obstacle. Fun begins when the new start point is not in free space, but is still inside the solid.

On most props, when a trace starts inside a world object, the trace doesn't go anywhere and  _~just_ spawns a subsequent trace.

But _sometimes_, on _some_ world objects (mostly map geometry), things behave differently.
When a subsequent trace starts inside a _special_ solid, it won't stop at the start point and will continue tracing until it hits another similar cursed object (or infinitely, idk). And that trace info will be polluted with the cursed object, so you can't even try to do dmg either.

For normal:
- :o: `-->[  o ] -->`

For cursed:
- :o: `-->[  ]  --->`
- :x: `-->[    --]-->`

Don't know how to fix TraceLine and if it is fixable or even if this is actually its intended behavior.
But there is `tr.fractionleftsolid`, so it ~should be~ is possible to recalculate starting point and redo the trace.
